File: boj/gold/boj-gold5/9663-N_Queens.py
# n개의 퀸을 배치 -> 무조건 모든 행에 퀸이 위치해야 함
# 0열 ~ n-1열까지
# 유망한지(promising)한지, 즉 이전의 열로 인해 영향을 받는지 검사하는 함수를 추가한다
def adjacent(x):
    for i in range(x):
        # 같은 행에 있으면 안됨 or 같은 대각선에 있으면 안됨
        if col[x] == col[i] or abs(col[x]-col[i]) == x-i:
            return False
    return True


def dfs(x):
    global result
    if x == n:
        result += 1
    else:
        for i in range(n):
            col[x] = i
            if adjacent(x):
                dfs(x+1)


n = int(input())
col = [0] * n
result = 0
dfs(0)
print(result)

File: boj/gold/boj-gold5/9663-N_Queens.cpp
#include <iostream>
#define MAX 15
using namespace std;


int col[MAX];
int n;
int total = 0;

bool adjacent(int x) {
    for (int i = 0; i < x; i++) {
        // 같은 행에 있으면 안됨 || 같은 대각선에 있으면 안됨
        if (col[x] == col[i] || abs(col[x]-col[i]) == x-i) {
            return false;
        }
    }
    return true;
}
void nqueens(int x) {
    if (x == n) total ++;
    else {
        for (int i=0; i < n; i++) {
            col[x] = i; // 해당 위치에 퀸을 위치시킨다
            if (adjacent(x)) nqueens(x+1);
        }
    }

}

int main() {
    cin >> n;
    nqueens(0);
    cout << total;
}